File: src/loco_MuJoCo/ros2/ros2_py/robot_controller.py
import sys
import select
import rclpy
from rclpy.node import Node
from my_interfaces.srv import RobotCtrl

# 全局变量（提前初始化，避免未绑定）
current_action_key = 'stop'
action_switch_flag = False


class RobotController(Node):
    def __init__(self, name):
        super().__init__(name)
        # 创建服务客户端，连接机器人控制服务
        self.client = self.create_client(RobotCtrl, 'robot_control')
        # 等待服务上线
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('等待服务 "robot_control" 启动...')
        self.get_logger().info("操作提示：w=行走 | s=下蹲 | q=退出 | 其他键=停止")

    def result_callback(self, future):
        """服务调用结果的回调函数"""

        response = future.result()
        self.get_logger().info(f'指令执行成功：{response.message}')


    def send_request(self, action):
        """发送控制指令到服务端"""

        request = RobotCtrl.Request()
        request.command = action  # 传递动作指令
        self.client.call_async(request).add_done_callback(self.result_callback)
        self.get_logger().info(f'已发送指令：{action}')



def read_keypress_non_blocking():
    """非阻塞读取终端按键"""
    if select.select([sys.stdin], [], [], 0) == ([sys.stdin], [], []):
        key = sys.stdin.read(1)
        return key.strip()
    return None


def main(args=None):
    # 关键：声明使用全局变量（修复核心错误）
    global current_action_key, action_switch_flag
    
    rclpy.init(args=args)
    robot_controller = RobotController("robot_controller")

    # 初始发送停止指令（此时current_action_key已绑定全局值'stop'）
    robot_controller.send_request(current_action_key)

    try:
        # 主循环：非阻塞读取按键 + 处理ROS事件
        while rclpy.ok():
            # 1. 读取终端按键
            key = read_keypress_non_blocking()
            if key:
                # 按键逻辑
                if key == 'w' and current_action_key != "walk":
                    current_action_key = "walk"
                    action_switch_flag = True
                elif key == 's' and current_action_key != "squat":
                    current_action_key = "squat"
                    action_switch_flag = True
                elif key == 'q':  # q键退出
                    robot_controller.get_logger().info("收到退出指令，程序终止")
                    break
                elif key not in ['w', 's']:  # 其他键停止
                    if current_action_key != "stop":
                        current_action_key = "stop"
                        action_switch_flag = True

            # 2. 检测动作切换并发送指令
            if action_switch_flag:
                robot_controller.send_request(current_action_key)
                action_switch_flag = False  # 重置标记

            # 3. 非阻塞处理ROS回调
            rclpy.spin_once(robot_controller, timeout_sec=0.01)

    except KeyboardInterrupt:
        robot_controller.get_logger().info("用户强制退出")
    finally:
        robot_controller.destroy_node()
        rclpy.shutdown()
        print("程序已退出")


if __name__ == '__main__':
    main()

File: src/humanoid_3d_animation/ros2/mujoco_ros2_cpp/robot_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_interfaces/srv/robot_ctrl.hpp"
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>
#include <string>

// 全局变量
std::string current_action_key = "stop";
bool action_switch_flag = false;

// 非阻塞键盘读取
char read_keypress_non_blocking() {
    struct termios oldt, newt;
    int ch;
    int oldf;

    // 获取终端设置
    tcgetattr(STDIN_FILENO, &oldt);
    newt = oldt;
    // 禁用 canonical 模式和回显
    newt.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &newt);
    // 设置非阻塞模式
    oldf = fcntl(STDIN_FILENO, F_GETFL, 0);
    fcntl(STDIN_FILENO, F_SETFL, oldf | O_NONBLOCK);

    ch = getchar();

    // 恢复终端设置
    tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
    fcntl(STDIN_FILENO, F_SETFL, oldf);

    return (ch != EOF) ? ch : 0;
}

class RobotController : public rclcpp::Node {
public:
    RobotController() : Node("robot_controller") {
        // 创建服务客户端
        client_ = this->create_client<my_interfaces::srv::RobotCtrl>("robot_control");
        
        // 等待服务可用
        while (!client_->wait_for_service(std::chrono::seconds(1))) {
            if (!rclcpp::ok()) {
                RCLCPP_ERROR(this->get_logger(), "等待服务时被中断");
                return;
            }
            RCLCPP_INFO(this->get_logger(), "等待服务 'robot_control' 启动...");
        }
        
        RCLCPP_INFO(this->get_logger(), "操作提示：w=行走 | s=下蹲 | q=退出 | 其他键=停止");
    }

    void send_request(const std::string &action) {
        auto request = std::make_shared<my_interfaces::srv::RobotCtrl::Request>();
        request->command = action;

        // 发送异步请求
        auto result_future = client_->async_send_request(
            request,
            std::bind(&RobotController::result_callback, this, std::placeholders::_1)
        );
        RCLCPP_INFO(this->get_logger(), "已发送指令：%s", action.c_str());
    }

private:
    rclcpp::Client<my_interfaces::srv::RobotCtrl>::SharedPtr client_;

    void result_callback(rclcpp::Client<my_interfaces::srv::RobotCtrl>::SharedFuture future) {
        auto response = future.get();
        RCLCPP_INFO(this->get_logger(), "指令执行成功：%s", response->message.c_str());
    }
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<RobotController>();

    // 初始发送停止指令
    node->send_request(current_action_key);

    try {
        while (rclcpp::ok()) {
            // 读取键盘输入
            char key = read_keypress_non_blocking();
            if (key) {
                if (key == 'w' && current_action_key != "walk") {
                    current_action_key = "walk";
                    action_switch_flag = true;
                } else if (key == 's' && current_action_key != "squat") {
                    current_action_key = "squat";
                    action_switch_flag = true;
                } else if (key == 'q') {  // 退出
                    RCLCPP_INFO(node->get_logger(), "收到退出指令，程序终止");
                    break;
                } else if (key != 'w' && key != 's') {  // 其他键停止
                    if (current_action_key != "stop") {
                        current_action_key = "stop";
                        action_switch_flag = true;
                    }
                }
            }

            // 发送指令
            if (action_switch_flag) {
                node->send_request(current_action_key);
                action_switch_flag = false;
            }

            // 处理ROS事件
            rclcpp::spin_some(node);
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }
    } catch (const std::exception &e) {
        RCLCPP_ERROR(node->get_logger(), "异常：%s", e.what());
    }

    rclcpp::shutdown();
    return 0;
}